kafka-ui-api/src/main/java/com/provectus/kafka/ui/emitter/AbstractEmitter.java

```diff
@@ -0,0 +1,75 @@
+package com.provectus.kafka.ui.emitter;
+
+import com.provectus.kafka.ui.model.TopicMessage;
+import com.provectus.kafka.ui.model.TopicMessageConsuming;
+import com.provectus.kafka.ui.model.TopicMessageEvent;
+import com.provectus.kafka.ui.model.TopicMessagePhase;
+import com.provectus.kafka.ui.serde.RecordSerDe;
+import com.provectus.kafka.ui.util.ClusterUtil;
+import java.time.Duration;
+import java.time.Instant;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.utils.Bytes;
+import reactor.core.publisher.FluxSink;
+
+public abstract class AbstractEmitter {
+  private static final Duration POLL_TIMEOUT_MS = Duration.ofMillis(1000L);
+
+  private final RecordSerDe recordDeserializer;
+  private long bytes = 0;
+  private int records = 0;
+  private long elapsed = 0;
+
+  public AbstractEmitter(RecordSerDe recordDeserializer) {
+    this.recordDeserializer = recordDeserializer;
+  }
+
+  protected ConsumerRecords<Bytes, Bytes> poll(
+      FluxSink<TopicMessageEvent> sink, Consumer<Bytes, Bytes> consumer) {
+    Instant start = Instant.now();
+    ConsumerRecords<Bytes, Bytes> records = consumer.poll(POLL_TIMEOUT_MS);
+    Instant finish = Instant.now();
+    sendConsuming(sink, records, Duration.between(start, finish).toMillis());
+    return records;
+  }
+
+  protected FluxSink<TopicMessageEvent> sendMessage(FluxSink<TopicMessageEvent> sink,
+                             ConsumerRecord<Bytes, Bytes> msg) {
+    final TopicMessage topicMessage = ClusterUtil.mapToTopicMessage(msg, recordDeserializer);
+    return sink.next(
+        new TopicMessageEvent()
+            .type(TopicMessageEvent.TypeEnum.MESSAGE)
+            .message(topicMessage)
+    );
+  }
+
+  protected void sendPhase(FluxSink<TopicMessageEvent> sink, String name) {
+    sink.next(
+        new TopicMessageEvent()
+          .type(TopicMessageEvent.TypeEnum.PHASE)
+          .phase(new TopicMessagePhase().name(name))
```

Comment:
what is the purpose of having this messages in UI? Debugging?

Reply:
To provide more interaction for end users